### PR TITLE
Use CultureAspect in Lucene

### LINKED ISSUE #7306: Fix CultureAspect extensibility
We need to refactor code in places where we access directly the LocalizationPart to get the ContentItem current culture.
Instead we should use the `_contentManager.PopulateAspectAsync(contentItem, new CultureAspect());` or something similar so that if someone implements a different ContentPart that could alter the ContentItem culture ; it evaluates both.

## TODO : 

- [ ] Evaluate changes to be made with CultureAspect class.
- [ ] Create a Unit Test that test extensibility.

## Examples : 

https://github.com/OrchardCMS/OrchardCore/blob/cfe0237eb843c103f113d11dfe093c0343ca572b/src/OrchardCore.Modules/OrchardCore.Lucene/Services/LuceneIndexingService.cs#L189

https://github.com/OrchardCMS/OrchardCore/blob/cfe0237eb843c103f113d11dfe093c0343ca572b/src/OrchardCore.Modules/OrchardCore.Lucene/Handler/LuceneIndexingContentHandler.cs#L85

So far it is the only 2 places I found where we do this. But we probably also use a similar technique with Recipes.
Please report other places where it needs to be fixed here.

### PR DESCRIPTION
**Pull request by hishamco**
Fixes #7306

### CODE REVIEW

**Comment by hishamco** [comment]
@Skrypt before I move further into all other places if there are, could you please let me know how to test the localization change in Lucene coz this is the first time I will try this module ;)

**Comment by Skrypt** [comment]
Create a custom Content Type.
Associate the LocalizationPart on it.
Create 2 different Lucene Index targetting 2 different specific cultures of your choice.
Create content items and check in which Index it gets indexed by using the match_all query.

**Comment by hishamco** [comment]
Oops, seems `indexSettings.Culture` is a string ;)

**Comment by Skrypt** [comment]
I think we added the default culture assignment to the CultureAspect because we presumed it would be used to get populated for the frontend only. We made a mistake by thinking so. I don't want to break the actual other implementations, so it will require some testing of Content Localization feature.

**Comment by hishamco** [comment]
> Here when we index we don't want this to happen because we want that if a Content Item has no LocalizationPart that the culture returned/evaluated/populated be null.

I see, I will try to consider this, but my question what will happen if the default culture return? Seems it 'll be indexed with that culture, Am I right? how this will affect lucene search?

**Comment by Skrypt** [comment]
It will get indexed in the wrong index 😉

**Comment by hishamco** [comment]
> We made a mistake by thinking so. I don't want to break the actual other implementations, so it will require some testing of Content Localization feature (if we do any refactoring).

I think we shouldn't break anything at this stage, probably we can look into Content Localization after `1.0.0`, IMHO for now we can use `CultureAspect` whenever it's possible. I will try to make tweak this to make it work in Lucene

**Comment by hishamco** [comment]
> Here when we index we don't want this to happen because we want that if a Content Item has no LocalizationPart that the culture returned/evaluated/populated be null.

@Skrypt is that also true if the index set for any culture?

**Comment by Skrypt** [comment]
Yes, it needs to be equal to what is stored in a LocalizationPart. If there is no LocalizationPart then we should index a null value.

**Comment by Skrypt** [comment]
Found what would work.

```C#
                                var contentItemCultureAspect = await contentManager.PopulateAspectAsync(context.ContentItem, new CultureAspect(){ Culture = null });
                                var ignoreIndexedCulture = settings.Culture == "any" ? false : contentItemCultureAspect.Culture?.Name != settings.Culture;

```

File: src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs
Comment (Skrypt) [inline src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs:27]: Here we end up with the same problem. We don't want to know about any part from here, it's the PopulateAspectAsync that should take care of returning the proper CultureAspect.

Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs:27]: Here in this Method we would need to be able to say default the value to CultureInfo.CurrentUICulture or not.

**Comment by hishamco** [comment]
@Skrypt frankly I already used `PopulateAspectAsync` before `GetContentCultureAsync()`, but I remembered @sebastienros suggest to use it long time back instead of `LocalizationPart`, which is make sense instead of repeating things, so fix it once and use it everywhere

If you complain about `contentItem.Has<LocalizationPart>` I already come up with a solution that works fine, which is to check if the retrieved cultures is supported or not

I suggest to continue with this PR unless the guys have another opinion

Thanks

File: src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs:27]: We could use `ILocalizationService` to ensure whether the `CultureAspect.Culture` is supported or not, this way no need to use `LocalizationPart` at all

**Comment by Skrypt** [comment]
I will take a look tomorrow. I don't have a preference for now. I agree though that if we could use the same method everywhere that be better. I tried fixing the GetContentCultureAsync method differently too but so far no luck. The only thing I don't want to see is the use of the LocalizationPart directly in these methods. The LocalizationPart should be affecting the result only when we do PopulateAspectAsync().

**Comment by hishamco** [comment]
> I tried fixing the GetContentCultureAsync method differently too but so far no luck. The only thing I don't want to see is the use of the LocalizationPart directly in these methods.

I already mentioned that above in my comment [here](https://github.com/OrchardCMS/OrchardCore/pull/7307#discussion_r507515995)

**Comment by Skrypt** [comment]
Yeah, sorry I'm tired. Going to sleep 😉

File: src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs
Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs:27]: I don't see what you mean. I need to check.

**Comment by hishamco** [comment]
No problem dude, sleep well, it's morning here ;)

File: src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs
Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs:27]: Good thing is that we're getting closer to something.

Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs:27]: Ok I can live with this for now but we need to find something better than looking for a specific ContentPart. Maybe more something like look for any ContentPart that inherits ILocalizable.

Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs:27]: Ok I can live with this for now but we need to find something better than looking for a specific ContentPart. Maybe more something like look for any ContentPart that inherits ILocalizable.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs:27]: The little issue is occur in the unit tests, coz the default culture will be supported, but I need to do more test to check what's going on

The reason is if I'm not wrong that `CultureInfo.CurrentUICulture` equals `CultureInfo.InstalledUICulture.Name`, Let us think while I finish some tests

Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs:27]: And also, I think it will break some Views where we look for a culture for defining RTL or LTR. Because now it could return a null value. That's mainly why I was not fixing it here.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.ContentLocalization.Abstractions/OrchardHelperExtensions.cs:27]: Did you check my comment [here](https://github.com/OrchardCMS/OrchardCore/pull/7307#discussion_r508064029)

**Comment by Skrypt** [comment]
While what Hisham did here in his global method looks fine it will break some UI features so we keep this on hold for more testing and fixing if we use this method. I had a simpler fix which doesn't break anything but @hishamco wants to fix the main method that returns the content item culture which now can return a null value which wasn't before.

**Review by Skrypt — CHANGES_REQUESTED** [review]
While what Hisham did here in his global method looks fine it will break some UI features so we keep this on hold for more testing and fixing if we use this method. I had a simpler fix which doesn't break anything but @hishamco wants to fix the main method that returns the content item culture which now can return a null value which wasn't before (this is what will break LocalizationPart UI views and else). 

This goes against the design of using the CultureAspect.Culture to be set to CultureInfo.CurrentUICulture by default.
Also, I believe that the CultureAspect was made to be used on frontend things and we return a default culture because a null culture doesn't exist for that matter. Though, here when we save the content item culture value it can be null. 

**We need to make a distinction between business logic and views/drivers logic.**

**Comment by hishamco** [comment]
> While what Hisham did here in his global method looks fine it will break some UI features

Give me some UI that was broken

> so we keep this on hold for more testing and fixing if we use this method

Agree, I will try to update this PR once I have time and have some extra tests

**Comment by Skrypt** [comment]
Try creating a Content Type that has no LocalizationPart. Create a content Item. Then add the LocalizationPart to it and try to edit it to see that it won't return a default culture to be saved against it. That's why we added the default culture on the CultureAspect I think.

**Comment by hishamco** [comment]
@Skrypt I have an idea probably crazy, but it solve the issue with minimal changes, what if we add a property something like `IsOriginalCultureNull` or any better name, with that the `Culture` property will still the same with default value `CultureInfo.CurrentUICulture`, but the new property will be useful especially for lucene to indicate that the original culture is `null`

**Comment by Skrypt** [comment]
Yes the GetContentItemCulture should have a param that says : here I want my default to be a null culture instead of the one provided by the CultureAspect.

**Comment by hishamco** [comment]
May be we can rename it to `GetContentCultureOrDefaultAsync()`

**Comment by Skrypt** [comment]
Nah I think the name is fine, just add a comment on the method.

**Comment by hishamco** [comment]
The comment is useful for those who have access to the code only, I see similar pattern in Microsoft APIs as well as OC

**Comment by Skrypt** [comment]
Comments on the method shows when you use the method elsewhere. You can add param comments. Of course inner comments won't show.

**Comment by hishamco** [comment]
The main question here, if we didn't add metadata in `CultureAspect` we will reply on `HasPart` which is the thing that you don't like, so your thought

**Comment by Skrypt** [comment]
We should not rely on any part. If you allow to use a different default culture, then you don't need to.

**Comment by hishamco** [comment]
> When you do populateAspect() it triggers the proper "LocalizationPartHandlers" that will affect the content item culture and it's fine like that. It abstract this part.

Got it, but we use

```csharp
var cultureAspect = await contentManager.PopulateAspectAsync(contentItem, new CultureAspect());
```

The culture will be `CultureInfo.CurrentUICulture` by default, so no change to know if the original culture is `null`, hope you got my point

**Comment by Skrypt** [comment]
I know this is quite difficult because it's a static class/method but something like this could work :  

``` C#
var cultureAspect = await contentManager.PopulateAspectAsync(contentItem, new CultureAspect(){ Culture = IsOriginalCultureNull ? null : CultureInfo.CurrentUICulture });
```

**Comment by hishamco** [comment]
I thought you will do it without `IsOriginalCultureNull` ;) ok, just leave it to me and I will see what can I come up with, I will keep you  updated

File: src/OrchardCore.Modules/OrchardCore.Lucene/Handler/LuceneIndexingContentHandler.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Lucene/Handler/LuceneIndexingContentHandler.cs:84]: Don't use the `IOrchardHelper` here. Use the `IContentManager.PopulateAspect`

The helper methods are for the ui/users of OC, not internal code.

**Review by deanmarcussen — COMMENTED** [review]
Reviewing this, I think the problem, as you have already noted @hishamco is that the `CultureAspect` might not be always useful, in that it returns a default.

In a mono culture site, or varied culture (as in some content is localized, some isn't) this isn't always useful.

Personally I think the `CultureAspect` should have a second property `HasCulture` returning a bool, set to false initially, whereby you can check first to see if a content item has had a culture specified.

This feels non breaking, and I think would resolve the issue in lucene, where it is necesary to know if there is a culture assigned or not, first, before then trying to get that culture.

There could be helper methods on the `IOrchardHelper` as well, but I would suggest to focus on getting the `CultureAspect` useful for all scenarios.

**Comment by hishamco** [comment]
IMHO we need to explicit assign `CultureInfo.CurrenUICulture` to `CultureAspect.Culture`, so `HasCulture` could be a ready only property, but for now it's fine, unless we know exactly where are the places that need to set culture by default

**Comment by hishamco** [comment]
@Skrypt could you please review this?

File: src/OrchardCore.Modules/OrchardCore.Lucene/Handler/LuceneIndexingContentHandler.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Lucene/Handler/LuceneIndexingContentHandler.cs:85]:
```suggestion
                    var ignoreIndexedCulture = indexSettings.Culture == "any" ? false : culture != indexSettings.Culture;
```

**Review by deanmarcussen — COMMENTED** [review]
I would wait for @Skrypt to check it, but I think you have a better / more useful aspect api now

**Comment by Skrypt** [comment]
I will review this soon. Looks good @hishamco

**Review by Skrypt — APPROVED** [review]

**Pull request merged**
